common/src/main/java/com/revenuecat/purchases/common/HTTPClient.kt

```diff
@@ -101,9 +114,53 @@ class HTTPClient(
             connection.disconnect()
         }
 
-        val responseBody = payload?.let { JSONObject(it) } ?: throw IOException("Network call payload is null.")
         log(LogIntent.DEBUG, NetworkStrings.API_REQUEST_COMPLETED.format(connection.requestMethod, path, responseCode))
-        return Result(responseCode, responseBody)
+        if (payload == null) {
+            throw IOException(NetworkStrings.HTTP_RESPONSE_PAYLOAD_NULL)
+        }
+
+        val resultFromBackend = HTTPResult(responseCode, payload)
+        connection.getETagHeader()?.let { eTagInResponse ->
+            if (eTagManager.shouldUseCachedVersion(responseCode)) {
+                val storedResult = eTagManager.getStoredResult(urlPathWithVersion)
+                return storedResult
+                    ?: if (refreshETag) {
+                        log(LogIntent.WARNING, NetworkStrings.ETAG_CALL_ALREADY_RETRIED.format(resultFromBackend))
+                        resultFromBackend
+                    } else {
+                        log(LogIntent.WARNING, NetworkStrings.ETAG_RETRYING_CALL)
+                        performRequest(path, body, authenticationHeaders, refreshETag = true)
+                    }
+            }
+
+            eTagManager.storeBackendResultIfNoError(urlPathWithVersion, resultFromBackend, eTagInResponse)
```

Comment:
I definitely get the urge to get this into eTag Manager instead of having 4 public (well, internal) methods for the class. It's a shame we didn't find a clean method or a clean way to achieve that. 
That said, I think it's fine to merge and then we can move it later if we come up with a better way

Reply:
I agree. What if I do the following:

In ETagManager:
```
   internal fun getHTTPResultFromCacheOrBackend(
        responseCode: Int,
        payload: String,
        connection: HttpURLConnection,
        urlPathWithVersion: String,
        refreshETag: Boolean
    ): HTTPResult? {
        val resultFromBackend = HTTPResult(responseCode, payload)
        connection.getETagHeader()?.let { eTagInResponse ->
            if (shouldUseCachedVersion(responseCode)) {
                val storedResult = getStoredResult(urlPathWithVersion)
                return storedResult
                    ?: if (refreshETag) {
                        log(LogIntent.WARNING, NetworkStrings.ETAG_CALL_ALREADY_RETRIED.format(resultFromBackend))
                        resultFromBackend
                    } else {
                      null  
                    }
            }

            storeBackendResultIfNoError(urlPathWithVersion, resultFromBackend, eTagInResponse)
        }
        return resultFromBackend
    }
```

And then in HTTPClient:
```
         val callResult = eTagManager.getHTTPResultFromCacheOrBackend(
            responseCode,
            payload,
            connection,
            urlPathWithVersion,
            refreshETag
        )
        if (callResult == null) {
            log(LogIntent.WARNING, NetworkStrings.ETAG_RETRYING_CALL)
            return performRequest(path, body, authenticationHeaders, refreshETag = true)
        }
        return callResult
```

Reply:
that looks pretty nice. It could be a follow-up PR if you'd like to get this merged for now

Reply:
I'll do a following PR I have almost ready :)